File: packages/js/core/src/uri-resolution/UriResolutionResult.ts
import { UriPackageOrWrapper, IUriResolutionStep } from ".";
import { Uri, IWrapPackage, Wrapper } from "..";

import { Result, ResultOk, ResultErr } from "@polywrap/result";

export class UriResolutionResult<TError = undefined> {
  public result: Result<UriPackageOrWrapper, TError>;
  public history?: IUriResolutionStep<unknown>[];

  static ok<TError = undefined>(uri: Uri): Result<UriPackageOrWrapper, TError>;
  static ok<TError = undefined>(
    uri: Uri,
    wrapPackage: IWrapPackage
  ): Result<UriPackageOrWrapper, TError>;
  static ok<TError = undefined>(
    uri: Uri,
    wrapper: Wrapper
  ): Result<UriPackageOrWrapper, TError>;
  static ok<TError = undefined>(
    uri: Uri,
    packageOrWrapper?: IWrapPackage | Wrapper
  ): Result<UriPackageOrWrapper, TError> {
    if (!packageOrWrapper) {
      return ResultOk({
        type: "uri",
        uri,
      } as UriPackageOrWrapper);
    }

    const wrapPackage = packageOrWrapper as Partial<IWrapPackage>;

    if (wrapPackage.createWrapper) {
      return ResultOk({
        type: "package",
        uri,
        package: wrapPackage as IWrapPackage,
      } as UriPackageOrWrapper);
    }

    const wrapper = packageOrWrapper as Partial<Wrapper>;

    if (wrapper.invoke) {
      return ResultOk({
        type: "wrapper",
        uri,
        wrapper: wrapper as Wrapper,
      } as UriPackageOrWrapper);
    }

    throw new Error("Unexpected type when creating UriResolutionResponse");
  }

  static err<TError = unknown>(
    error: TError
  ): Result<UriPackageOrWrapper, TError> {
    return ResultErr(error);
  }
}

File: packages/py/polywrap-core/polywrap_core/uri_resolution/uri_resolution_result.py
from typing import List, Optional

from result import Err, Ok, Result

from ..types import IUriResolutionStep, IWasmPackage, Uri, UriPackage, UriPackageOrWrapper, UriWrapper, Wrapper


class UriResolutionResult:
    result: Result[UriPackageOrWrapper, Exception]
    history: Optional[List[IUriResolutionStep]]

    @staticmethod
    def ok(
        uri: Uri, package: Optional[IWasmPackage] = None, wrapper: Optional[Wrapper] = None
    ) -> Result[UriPackageOrWrapper, Exception]:
        if wrapper:
            return Ok(UriWrapper(uri=uri, wrapper=wrapper))
        elif package:
            return Ok(UriPackage(uri=uri, package=package))
        else:
            return Ok(uri)

    @staticmethod
    def err(error: Exception) -> Result[UriPackageOrWrapper, Exception]:
        return Err(error)
